File: 00070-climbing-stairs/main.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef pair<int, int> pi;
#define fi first
#define se second
#define endl '\n'
#define debug(x) cout << #x << ": " << x << endl
#define REP(i, a, b) for (int i = a; i < b; i++)

template <unsigned int N>
struct Fibonacci {
    static constexpr unsigned int value = Fibonacci<N-1>::value + Fibonacci<N-2>::value;
};

template <>
struct Fibonacci<0u> {
    static constexpr unsigned int value = 0;
};

template <>
struct Fibonacci<1u> {
    static constexpr unsigned int value = 1;
};

template <unsigned int N>
constexpr unsigned int fib() {
    return Fibonacci<N>::value;
}

template<unsigned int N>
constexpr std::array<unsigned int, N> dp = []() constexpr {
    std::array<unsigned int, N> a;
    for (unsigned int i = 0; i < N; i++) {
        a[i] = fib<i>();
    }
    return a;
}();

class Solution {
public:
    int climbStairs(int n) {
        return dp<46>[n];
    }
};

signed main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    Solution sol;
    auto ans = sol.climbStairs(11);

}

File: 00070-climbing-stairs/main.py
def factorial(b):
    if b == 0:
        return 1
    return factorial(b - 1)*b


def non_repeating(a, b):
    if (b > a):
        a, b = b, a
    prod = 1
    for i in range(a + 1, a + b + 1):
        prod *= i
    return prod/factorial(b)


class Solution:
    def climbStairs(self, n: int) -> int:
        total = 0
        for i in range(n//2 + 1):
            a, b = i, n - 2*i
            non_rep = non_repeating(a, b)
            total += non_rep
        return int(total)


sol = Solution()
ans = sol.climbStairs(11)
print("ans", ans)
